src/Npgsql/NpgsqlConnection.cs

```diff
@@ -198,21 +194,23 @@ void GetPoolAndSettings()
                 }
 
                 // Maybe pooling is off
-                if (!Settings.Pooling)
-                    return;
-
-                // Connstring may be equivalent to one that has already been seen though (e.g. different
-                // ordering). Have NpgsqlConnectionStringBuilder produce a canonical string representation
-                // and recheck.
-                var canonical = Settings.ConnectionString;
-                if (pools.TryGetValue(canonical, out _pool))
-                    pools[_connectionString] = _pool;
-                else
+                if (Settings.Pooling)
                 {
-                    // Really unseen, need to create a new pool
-                    _pool = pools[_connectionString] = new ConnectorPool(Settings, canonical);
-                    if (_connectionString != canonical)
-                        pools[canonical] = _pool;
+                    // Connstring may be equivalent to one that has already been seen though (e.g. different
+                    // ordering). Have NpgsqlConnectionStringBuilder produce a canonical string representation
+                    // and recheck.
+                    var canonical = Settings.ConnectionString;
+
+                    if (pools.TryGetValue(canonical, out _pool))
+                        _pool = pools.GetOrAdd(_connectionString, _pool); // Assign the connection string the canonical pool. If someone beat us to it use whatever is there already.
+                    else
+                    {
+                        // Really unseen, need to create a new pool
+                        // The canonical pool is the 'base' pool so we need to set that up first. If someone beats us to it use what they put.
+                        // The connection string pool can either be added here or above, if it's added above we should just use that.
+                        _pool = pools.GetOrAdd(canonical, new ConnectorPool(Settings, canonical));
```

Comment:
There's a subtle bug here. If a race condition occurs, it's possible for two instances of `ConnectorPool` to be instantiated, and one of them to get discarded. That's not a problem at all (it's how things work with ConcurrentDictionary), except that the constructor for `ConnectorPool` increments the counter for the number of active pools... So we end up increasing it by two (or even more) although only one new pool is created.

This can be fixed by checking whether the result of `GetOrAdd()` is reference-equals the instance we created - it would tell us if our instance is the one that made it into the ConcurrentDictionary. If this is the case, we can increment the counter. This would move the counter update code out of the constructor.